File: frontend/src/services/bank-account-service.ts
import api from './api';
import type {
  BankAccount,
  CreateBankAccount,
  UpdateBankAccount,
  BankTransaction,
  CreateBankTransaction,
} from '@/types';

export class BankAccountService {
  private static readonly BASE_PATH = '/api/v1/bank-accounts';

  // Get all bank accounts
  static async getAll(): Promise<BankAccount[]> {
    const response = await api.get<BankAccount[]>(this.BASE_PATH);
    return response.data;
  }

  // Get bank account by ID
  static async getById(id: string): Promise<BankAccount> {
    const response = await api.get<BankAccount>(`${this.BASE_PATH}/${id}`);
    return response.data;
  }

  // Get bank accounts by pump master ID
  static async getByPumpMasterId(pumpMasterId: string): Promise<BankAccount[]> {
    const response = await api.get<BankAccount[]>(
      `${this.BASE_PATH}/pump/${pumpMasterId}`
    );
    return response.data;
  }

  // Search bank accounts with filters
  static async search(params: {
    accountHolderName?: string;
    accountNumber?: string;
    bank?: string;
    pumpMasterId?: string;
  }): Promise<BankAccount[]> {
    const queryParams = new URLSearchParams();
    if (params.accountHolderName)
      queryParams.append('accountHolderName', params.accountHolderName);
    if (params.accountNumber)
      queryParams.append('accountNumber', params.accountNumber);
    if (params.bank) queryParams.append('bank', params.bank);
    if (params.pumpMasterId)
      queryParams.append('pumpMasterId', params.pumpMasterId);

    const response = await api.get<BankAccount[]>(
      `${this.BASE_PATH}/search?${queryParams}`
    );
    return response.data;
  }

  // Create new bank account
  static async create(bankAccount: CreateBankAccount): Promise<BankAccount> {
    const response = await api.post<BankAccount>(this.BASE_PATH, bankAccount);
    return response.data;
  }

  // Update existing bank account
  static async update(
    id: string,
    bankAccount: UpdateBankAccount
  ): Promise<BankAccount> {
    const response = await api.put<BankAccount>(
      `${this.BASE_PATH}/${id}`,
      bankAccount
    );
    return response.data;
  }

  // Delete bank account
  static async delete(id: string): Promise<void> {
    await api.delete(`${this.BASE_PATH}/${id}`);
  }

  // Get transactions for a bank account
  static async getTransactions(
    bankAccountId: string
  ): Promise<BankTransaction[]> {
    const response = await api.get<BankTransaction[]>(
      `${this.BASE_PATH}/${bankAccountId}/transactions`
    );
    return response.data;
  }

  // Create credit transaction
  static async credit(
    bankAccountId: string,
    transaction: CreateBankTransaction
  ): Promise<BankTransaction> {
    const response = await api.post<BankTransaction>(
      `${this.BASE_PATH}/${bankAccountId}/credit`,
      transaction
    );
    return response.data;
  }

  // Create debit transaction
  static async debit(
    bankAccountId: string,
    transaction: CreateBankTransaction
  ): Promise<BankTransaction> {
    const response = await api.post<BankTransaction>(
      `${this.BASE_PATH}/${bankAccountId}/debit`,
      transaction
    );
    return response.data;
  }
}

File: backend/src/main/java/com/reallink/pump/services/BankAccountService.java
package com.reallink.pump.services;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.validation.annotation.Validated;

import com.reallink.pump.dto.request.CreateBankAccountRequest;
import com.reallink.pump.dto.request.UpdateBankAccountRequest;
import com.reallink.pump.dto.response.BankAccountResponse;
import com.reallink.pump.entities.BankAccount;
import com.reallink.pump.entities.DailyClosingBalance;
import com.reallink.pump.entities.PumpInfoMaster;
import com.reallink.pump.exception.PumpBusinessException;
import com.reallink.pump.mapper.BankAccountMapper;
import com.reallink.pump.repositories.BankAccountRepository;
import com.reallink.pump.repositories.DailyClosingBalanceRepository;
import com.reallink.pump.repositories.PumpInfoMasterRepository;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import lombok.RequiredArgsConstructor;

@Service
@Validated
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class BankAccountService {

    private final BankAccountRepository repository;
    private final PumpInfoMasterRepository pumpInfoMasterRepository;
    private final BankAccountMapper mapper;
    private final BankTransactionService bankTransactionService;
    private final DailyClosingBalanceRepository dailyClosingBalanceRepository;

    public List<BankAccountResponse> getAll() {
        return repository.findAll().stream()
                .map(bankAccount -> {
                    BankAccountResponse response = mapper.toResponse(bankAccount);
                    setCurrentBalance(response, bankAccount);
                    return response;
                })
                .toList();
    }

    public Page<BankAccountResponse> getAllPaginated(Pageable pageable) {
        return repository.findAll(pageable)
                .map(mapper::toResponse);
    }

    public BankAccountResponse getById(@NotNull UUID id) {
        BankAccount bankAccount = repository.findById(id).orElse(null);
        if (bankAccount == null) {
            throw new PumpBusinessException("BANK_ACCOUNT_NOT_FOUND", "Bank account with ID " + id + " not found");
        }
        BankAccountResponse response = mapper.toResponse(bankAccount);
        setCurrentBalance(response, bankAccount);
        return response;
    }

    public List<BankAccountResponse> getByPumpMasterId(@NotNull UUID pumpMasterId) {
        return repository.findByPumpMaster_Id(pumpMasterId).stream()
                .map(mapper::toResponse)
                .toList();
    }

    public BankAccountResponse getByAccountNumberAndPumpMasterId(@NotNull String accountNumber, @NotNull UUID pumpMasterId) {
        BankAccount bankAccount = repository.findByAccountNumberAndPumpMaster_Id(accountNumber, pumpMasterId).orElse(null);
        if (bankAccount == null) {
            throw new PumpBusinessException("BANK_ACCOUNT_NOT_FOUND",
                    "Bank account with number '" + accountNumber + "' and pump master ID " + pumpMasterId + " not found");
        }
        return mapper.toResponse(bankAccount);
    }

    public List<BankAccountResponse> getByBankAndPumpMasterId(@NotNull String bank, @NotNull UUID pumpMasterId) {
        return repository.findByBankAndPumpMaster_Id(bank, pumpMasterId).stream()
                .map(mapper::toResponse)
                .toList();
    }

    public List<BankAccountResponse> getByIfscCodeAndPumpMasterId(@NotNull String ifscCode, @NotNull UUID pumpMasterId) {
        return repository.findByIfscCodeAndPumpMaster_Id(ifscCode, pumpMasterId).stream()
                .map(mapper::toResponse)
                .toList();
    }

    public List<BankAccountResponse> searchBankAccounts(String accountHolderName, String accountNumber,
            String ifscCode, String bank, String branch, UUID pumpMasterId) {
        return repository
                .findBySearchCriteria(accountHolderName, accountNumber, ifscCode, bank, branch, pumpMasterId)
                .stream()
                .map(mapper::toResponse)
                .toList();
    }

    public List<BankAccountResponse> getByAccountHolderNameContaining(@NotNull String accountHolderName) {
        return repository.findByAccountHolderNameContainingIgnoreCase(accountHolderName).stream()
                .map(mapper::toResponse)
                .toList();
    }

    @Transactional
    public BankAccountResponse create(@Valid CreateBankAccountRequest request) {
        // Check for duplicate account number
        if (repository.existsByAccountNumberAndPumpMaster_Id(request.getAccountNumber(), request.getPumpMasterId())) {
            throw new PumpBusinessException("DUPLICATE_BANK_ACCOUNT",
                    "Bank account with number '" + request.getAccountNumber() + "' already exists for this pump master");
        }
        // Fetch the PumpInfoMaster entity using pumpMasterId from the request
        PumpInfoMaster pumpMaster = pumpInfoMasterRepository.findById(request.getPumpMasterId()).orElse(null);
        if (pumpMaster == null) {
            throw new PumpBusinessException("INVALID_PUMP_MASTER",
                    "Pump master with ID " + request.getPumpMasterId() + " does not exist");
        }
        BankAccount bankAccount = mapper.toEntity(request);
        bankAccount.setPumpMaster(pumpMaster);
        BankAccount savedBankAccount = repository.save(bankAccount);
        BankAccountResponse response = mapper.toResponse(savedBankAccount);
        setCurrentBalance(response, savedBankAccount);
        return response;
    }

    @Transactional
    public BankAccountResponse update(@NotNull UUID id, @Valid UpdateBankAccountRequest request) {
        BankAccount existingBankAccount = repository.findById(id).orElse(null);
        if (existingBankAccount == null) {
            throw new PumpBusinessException("BANK_ACCOUNT_NOT_FOUND", "Bank account with ID " + id + " not found");
        }

        // Check for duplicate account number if number is being updated
        if (request.getAccountNumber() != null
                && !request.getAccountNumber().equals(existingBankAccount.getAccountNumber())
                && repository.existsByAccountNumberAndPumpMaster_IdAndIdNot(request.getAccountNumber(),
                        existingBankAccount.getPumpMaster().getId(), id)) {
            throw new PumpBusinessException("DUPLICATE_BANK_ACCOUNT",
                    "Bank account with number '" + request.getAccountNumber() + "' already exists for this pump master");
        }

        mapper.updateEntityFromRequest(request, existingBankAccount);
        BankAccount updatedBankAccount = repository.save(existingBankAccount);
        BankAccountResponse response = mapper.toResponse(updatedBankAccount);
        setCurrentBalance(response, updatedBankAccount);
        return response;
    }

    @Transactional
    public void delete(@NotNull UUID id) {
        if (!repository.existsById(id)) {
            throw new PumpBusinessException("BANK_ACCOUNT_NOT_FOUND", "Bank account with ID " + id + " not found");
        }
        repository.deleteById(id);
    }

    public boolean existsByAccountNumberAndPumpMasterId(@NotNull String accountNumber, @NotNull UUID pumpMasterId) {
        return repository.existsByAccountNumberAndPumpMaster_Id(accountNumber, pumpMasterId);
    }

    public long getCountByPumpMasterId(@NotNull UUID pumpMasterId) {
        return repository.countByPumpMasterId(pumpMasterId);
    }

    private void setCurrentBalance(BankAccountResponse response, BankAccount bankAccount) {
        Optional<DailyClosingBalance> latestBalance = dailyClosingBalanceRepository.findLatestByBankAccountIdAndDateBeforeOrEqual(bankAccount.getId(), LocalDate.now());
        if (latestBalance.isPresent()) {
            response.setCurrentBalance(latestBalance.get().getClosingBalance());
        } else {
            BigDecimal transactionSum = bankTransactionService.getBalanceByBankAccountId(bankAccount.getId());
            response.setCurrentBalance(bankAccount.getOpeningBalance().add(transactionSum));
        }
    }
}
